base/broadcast.jl

```diff
@@ -219,6 +219,12 @@ argtype(bc::Broadcasted) = argtype(typeof(bc))
 _eachindex(t::Tuple{Any}) = t[1]
 _eachindex(t::Tuple) = CartesianIndices(t)
 
+Base.IndexStyle(bc::Broadcasted) = IndexStyle(typeof(bc))
+Base.IndexStyle(::Type{<:Broadcasted{<:Any,Tuple{Axis}}}) where {Axis} = IndexStyle(Axis)
+Base.IndexStyle(::Type{<:Broadcasted{<:Any}}) = IndexCartesian()
+
+Base.LinearIndices(bc::Broadcasted) = LinearIndices(eachindex(bc))
```

Comment:
```suggestion
```

We cannot define `LinearIndices` for `Broadcasted`s since they don't support linear indexing. They're not `<: AbstractArray`, so they don't the fancy linear->cartesian methods that do this conversion.  Did you need this to make any of the reductions work?

Reply:
Yes, we need it for this `_mapreduce` to work:

https://github.com/JuliaLang/julia/blob/85f6405c8d81a2de2663034861e10a5668beddbd/base/reduce.jl#L305-L306

Why `Broadcasted`s "don't support linear indexing"?  I thought it's fine because we get to this `_mapreduce` implementation only if `IndexStyle(::Type{<:Broadcasted})` returns `IndexLinear()`.  I thought `IndexStyle(x) == IndexLinear()` implies that `LinearIndices(x)` works.  Or am I misunderstanding the relationship between `IndexLinear` and `LinearIndices`?

Reply:
We've not very carefully laid out all the restrictions on what these functions mean for non-`AbstractArray` types, but my rule would be that:

```
A == A[LinearIndices(A)] == A[CartesianIndices(A)]
```

Defining this for all Braodcasteds would totally break that.  If we just need this for `IndexLinear` ones, then lets define it as:

```suggestion
Base.LinearIndices(bc::Broadcasted{<:Any,Tuple{Any}}) = axes(bc)[1]
```

This way it'd be either an outright error or indexing with the result will work.

Reply:
Ah, yes, thanks, this definition makes sense. It was wrong to define `LinearIndices` _for all_ `Broadcasted` instances.